pkg/controller/cluster/cluster_controller.go

```diff
@@ -119,18 +141,47 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 	if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {
 		// no-op if finalizer has been removed.
 		if !util.Contains(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer) {
-			klog.Infof("reconciling cluster object %v causes a no-op as there is no finalizer.", name)
+			klog.Infof("Reconciling cluster object %v causes a no-op as there is no finalizer.", name)
 			return reconcile.Result{}, nil
 		}
 
-		klog.Infof("reconciling cluster object %v triggers delete.", name)
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, errors.WithStack(err)
+		}
+
+		if len(children) > 0 {
+			klog.Infof("Deleting cluster %s: %d children still exist, will requeue", name, len(children))
+			for _, child := range children {
+				accessor, err := meta.Accessor(child)
+				if err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "couldn't create accessor for %T", child)
```

Comment:
Should we aggregate all errors in this for loop so we can try to delete as much as we can?

Reply:
The deletion happens in a later step, I think Accessor is unlikely to fail. But it might be a good idea to aggregate the errors for the actual deletion loop

Reply:
I agree that Accessor is unlikely to fail, but I'd still like to include it in errList instead of returning early.

Reply:
I think this is my only remaining comment